torch/_dynamo/guards.py

```diff
@@ -75,6 +75,16 @@
             torch.are_deterministic_algorithms_enabled,
         ),
         ("___is_torch_function_enabled", torch._C._is_torch_function_enabled),
+        (
+            "___current_backend",
+            lambda: torch._dynamo.eval_frame.guarded_backend_cache.most_recent_backend,
+        ),
+        (
+            "___lookup_backend",
+            lambda backend_obj_id: torch._dynamo.eval_frame.guarded_backend_cache.cached_backends[
```

Comment:
Can you explain why all of this code exists? It is not obvious to me. Naively, I expect a backend object to be a plain singleton object which I do not mind if it is kept live by cache entries, and I can just hold onto it directly and do an `is` test on it.

Reply:
Ok! We want to use "==" to compare the backend instead of "is" because 1. every time we call torch.compile(backend="inductor"), we actually creates a new _TorchCompileInductorWrapper object (here: https://github.com/pytorch/pytorch/blob/main/torch/__init__.py#L1715). 2. The previous _different_backend check uses "==" instead of "is", which I think it's reasonable since "is" cannot detect obj mutations. 


When we **check** the BACKEND_MATCH guard, we want to compare current backend with the backend associated with current cache. That backend should be baked in the check function of the cache as a constant when we **compiling** the BACKEND_MATCH guard. This pr baked in the id(backend) into the check function. We then need the cached_backends dictionary to map the id back to the original backend object so that we can compare with "==".

Reply:
(1) _TorchCompileInductorWrapper already has a defined `__eq__` it does not seem a stretch to just have it define `__hash__` and not hack around it like this, and move over to using `is`?

Reply:
Thanks for the comment! I considered the alternative of implementing the guard check function based on '__hash__', but  '__hash__' doesn't prevent collisions in its spirit: even If the hash of the backend happens to be the same we're not certain that these two backend will do the same thing.

'is' doesn't check '__eq__' or '__hash__', it just checks whether these two are the same object in memory. Using 'is' will have the lambda and ephemeral wrapper problem that @jansel mentions.